bindings/sundials4py/idas/idas_usersupplied.hpp

```diff
@@ -0,0 +1,497 @@
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ * SUNDIALS Copyright End
+ *----------------------------------------------------------------------------*/
+
+#ifndef _SUNDIALS4PY_IDAS_USERSUPPLIED_HPP
+#define _SUNDIALS4PY_IDAS_USERSUPPLIED_HPP
+
+#include <idas/idas.h>
+#include <idas/idas_ls.h>
+
+#include <sundials/sundials_core.hpp>
+
+#include "sundials4py_helpers.hpp"
+
+///////////////////////////////////////////////////////////////////////////////
+// IDAS user-supplied function table
+// Every integrator-level user-supplied function must be in this table.
+// The user-supplied function table is passed to IDAS as user_data.
+///////////////////////////////////////////////////////////////////////////////
+
+struct idas_user_supplied_fn_table
+{
+  // user-supplied function pointers
+  nb::object res, rootfn, ewtn, rwtn, resNLS;
+
+  // idas_ls user-supplied function pointers
+  nb::object lsjacfn, lsprecsetupfn, lsprecsolvefn, lsjactimessetupfn,
+    lsjactimesvecfn, lsjacresfn;
+
+  // idas quadrature user-supplied function pointers
+  nanobind::object resQ, resQS;
+
+  // idas FSA user-supplied function pointers
+  nanobind::object resS;
+};
+
+inline idas_user_supplied_fn_table* idas_user_supplied_fn_table_alloc()
+{
+  // We must use malloc since IDAFree calls free
+  auto fn_table = static_cast<idas_user_supplied_fn_table*>(
+    std::malloc(sizeof(idas_user_supplied_fn_table)));
+
+  // Zero out the memory
+  std::memset(fn_table, 0, sizeof(idas_user_supplied_fn_table));
+
+  return fn_table;
+}
+
+inline idas_user_supplied_fn_table* get_idas_fn_table(void* ida_mem)
+{
+  auto mem      = static_cast<IDAMem>(ida_mem);
+  auto fn_table = static_cast<idas_user_supplied_fn_table*>(mem->python);
+  if (!fn_table)
+  {
+    throw sundials4py::null_function_table(
+      "Failed to get Python function table from IDAS memory");
+  }
+  return fn_table;
+}
+
+///////////////////////////////////////////////////////////////////////////////
+// IDAS user-supplied functions
+///////////////////////////////////////////////////////////////////////////////
+
+template<typename... Args>
+inline int idas_res_wrapper(Args... args)
+{
+  return sundials4py::user_supplied_fn_caller<
+    std::remove_pointer_t<IDAResFn>, idas_user_supplied_fn_table, IDAMem,
+    1>(&idas_user_supplied_fn_table::res, std::forward<Args>(args)...);
+}
+
+using IDARootStdFn = int(sunrealtype t, N_Vector y, N_Vector yp,
+                         sundials4py::Array1d gout, void* user_data);
+
+inline int idas_rootfn_wrapper(sunrealtype t, N_Vector y, N_Vector yp,
+                               sunrealtype* gout_1d, void* user_data)
+{
+  auto ida_mem  = static_cast<IDAMem>(user_data);
+  auto fn_table = get_idas_fn_table(user_data);
+  auto fn       = nb::cast<std::function<IDARootStdFn>>(fn_table->rootfn);
+  auto nrtfn    = ida_mem->ida_nrtfn;
+
+  sundials4py::Array1d gout(gout_1d, {static_cast<unsigned long>(nrtfn)},
```

Comment:
```suggestion
  sundials4py::Array1d gout(gout_1d, {static_cast<size_t>(nrtfn)},
```
https://nanobind.readthedocs.io/en/latest/api_extra.html#_CPPv4N8nanobind7ndarray7ndarrayE7VoidPtr6size_tPK6size_t6handlePK7int64_tN6dlpack5dtypeEiic